controllers/secret_controller_test.go

```diff
@@ -0,0 +1,347 @@
+
+// wrongClusterIgnoreCopy tests that the secrets are not copied if written to the target cluster instead of local cluster
+func wrongClusterIgnoreCopy(t *testing.T, ctx context.Context, f *framework.Framework, namespace string) {
+	require := require.New(t)
+
+	err := f.Client.Create(ctx, generateReplicatedSecret(namespace))
+	require.NoError(err, "failed to create replicated secret to main cluster")
+
+	generatedSecrets := generateSecrets(namespace)
+
+	for _, s := range generatedSecrets {
+		err := testEnv.Clients[targetCopyToCluster].Create(ctx, s)
+		require.NoError(err, "failed to create secret to main cluster")
+	}
+
+	t.Log("check that the secrets were not copied to other cluster(s)")
+	for _, s := range generatedSecrets {
+		require.Never(func() bool {
+			return verifySecretCopied(t, ctx, targetCopyToCluster, s, nil)
+		}, timeout, interval)
+	}
+}
+
+func generateSecrets(namespace string) []*corev1.Secret {
+	return []*corev1.Secret{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      "test-secret-first",
+				Labels: map[string]string{
+					"secret-controller": "test",
+				},
+			},
+			Type: "Opaque",
+			StringData: map[string]string{
+				"key": "value",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      "test-secret-second",
+				Labels: map[string]string{
+					"secret-controller": "nomatch",
+				},
+			},
+			Type: "Opaque",
+			StringData: map[string]string{
+				"key": "value",
+			},
+		},
+	}
+}
+
+func generateReplicatedSecret(namespace string) *api.ReplicatedSecret {
+	return &api.ReplicatedSecret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      "fetch-secrets",
+		},
+		Spec: api.ReplicatedSecretSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{"secret-controller": "test"},
+			},
+			ReplicationTargets: []api.ReplicationTarget{
+				{
+					K8sContextName: "cluster-1",
+				},
+			},
+		},
+	}
+}
+
+// verifySecretCopied checks that the same key is copied to other clusters
+func verifySecretCopied(t *testing.T, ctx context.Context, origCluster string, originalSecret *corev1.Secret, verify func(*testing.T, *corev1.Secret, *corev1.Secret)) bool {
+
+	secretKey := types.NamespacedName{Namespace: originalSecret.Namespace, Name: originalSecret.Name}
+	for clusterKey, testClient := range testEnv.Clients {
+		if clusterKey == origCluster {
+			continue
+		}
+		secretCopy := &corev1.Secret{}
+		err := testClient.Get(ctx, secretKey, secretCopy)
+		if err != nil {
+			// All errors are false
+			return false
+		}
+
+		if verify != nil {
+			verify(t, originalSecret, secretCopy)
+		}
+	}
+	// In a single cluster setup, no copies is fine
+	return true
+}
+
+// verifySecretsMatch checks that the same secret is copied to other clusters
+func verifySecretsMatch(t *testing.T, ctx context.Context, localClient client.Client, remoteClusters []string, secrets map[string]struct{}, namespace string) bool {
```

Comment:
`verifySecretsMatch` checks that a secret having the same name exists and that the resource hash matches that of the source cluster. The resource hash is computer over the `Data` field. Maybe it is already done elsewhere but I think it would be good to also verify that metadata fields have expected values, namely labels and annotations. I suppose it couldn't hurt to also make sure that the owner reference is not set.

Reply:
This function is just for testing.

Reply:
But yes, if the tests would add more annotations (they don't currently), then that would. The lack of resource hash annotation (or wrong) is actually tested by another process in this test, although it's not necessarily visible without understanding how the copying and comparisons work.